segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java

```diff
@@ -492,61 +522,80 @@ void completeMerge(SegmentMetadata sourceMetadata) {
-            this.lastAppendedOffset = lastOffset;
-        }
-    }
-
     private void insert(long offset, ByteArraySegment data) {
         log.debug("{}: Insert (Offset = {}, Length = {}).", this.traceObjectId, offset, data.getLength());
 
         // There is a very small chance we might be adding data twice, if we get two concurrent requests that slipped past
         // the StorageReadManager. Fixing it would be complicated, so let's see if it poses any problems.
-        CacheIndexEntry entry = new CacheIndexEntry(offset, data.getLength());
-        long lastOffset = entry.getLastStreamSegmentOffset();
-        Exceptions.checkArgument(lastOffset < this.metadata.getStorageLength(), "entry",
-                "The given range of bytes (%d-%d) does not correspond to the StreamSegment range that is in Storage (%d).",
-                entry.getStreamSegmentOffset(), lastOffset, this.metadata.getStorageLength());
+        Exceptions.checkArgument(offset + data.getLength() <= this.metadata.getStorageLength(), "entry",
+                "The given range of bytes (Offset=%s, Length=%s) does not correspond to the StreamSegment range that is in Storage (%s).",
+                offset, data.getLength(), this.metadata.getStorageLength());
+
+        addToCacheAndIndex(data, offset, "Insert");
+    }
+
+    private CacheIndexEntry addToCacheAndIndex(BufferView data, long offset, String operationName) {
+        if (data.getLength() <= this.cacheStorage.getMaxEntryLength()) {
+            // The entire buffer can fit into one entry.
+            return addSingleEntryToCacheAndIndex(data, offset, operationName);
+        } else {
+            // Need to split the buffer into smaller entries and insert them individually.
+            int bufferOffset = 0;
+            CacheIndexEntry lastEntry = null;
+            while (bufferOffset < data.getLength()) {
+                int partLength = Math.min(data.getLength() - bufferOffset, this.cacheStorage.getMaxEntryLength());
+                lastEntry = addSingleEntryToCacheAndIndex(data.slice(bufferOffset, partLength), offset + bufferOffset, operationName);
+                bufferOffset += partLength;
+            }
+
+            return lastEntry;
+        }
+    }
+
+    private CacheIndexEntry addSingleEntryToCacheAndIndex(BufferView data, long offset, String operationName) {
+        int dataAddress = this.cacheStorage.insert(data);
+        CacheIndexEntry newEntry;
         ReadIndexEntry oldEntry;
-        synchronized (this.lock) {
-            this.cache.insert(getCacheKey(entry), data);
-            oldEntry = addToIndex(entry);
+        try {
+            newEntry = new CacheIndexEntry(offset, data.getLength(), dataAddress);
+            synchronized (this.lock) {
+                oldEntry = addToIndex(newEntry);
+            }
+        } catch (Throwable ex) {
```

Comment:
Is this a possible/expected case?

Reply:
No, not expected, but there can always be a random `RuntimException` being thrown from one of the sub-methods called from there. In this case, I want to make sure that the data we have just inserted into the CacheStorage will be removed, otherwise we'll start leaking.